Fix broadcasting via observed and dims

ricardoV94 opened the pull request:
Fixes #5993 

This PR also removes the support of ellipsis in dims and shape. This feature was not very advertised so hopefully this won't cause too much pain.

These tests illustrate the bug fix

```python
    def test_broadcast_by_dims(self):
        with pm.Model(coords={"broadcast_dim": range(3)}) as m:
            x = pm.Normal("x", mu=np.zeros((1,)), dims=("broadcast_dim",))
            assert x.eval().shape == (3,)

    def test_broadcast_by_observed(self):
        with pm.Model() as m:
            x = pm.Normal("x", mu=np.zeros((1,)), observed=np.zeros((3,)))
            assert x.eval().shape == (3,)
```

**Checklist**
+ [x] Explain important implementation details 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [x] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Major / Breaking Changes
- When shape is not provided, dims or observed are used to define the shape of a variable. If you want a variable that has dims or is observed to automatically resize when its inputs change, you must specify that explicitly via the shape argument. For example, `pm.Normal("likelihood", mu=mu, sigma=sigma, observed=data, dims="data", shape=mu.shape)`
- Remove support of Ellipsis (`...`) in shape and dims

## Bugfixes / New features
- Fix bug where distribution shape would not be broadcasted by dims or observed
- Allow specifying dims on the fly from observed
- Do not show shape-related dependencies of RandomVariables in `model_graph`

## Docs / Maintenance
- Update `_make_nice_attr_error` to suggest `pm.draw` instead of `.eval()`
- Add `_make_nice_attr_error` to `SymbolicDistribution`s

ricardoV94 committed:
Remove ellipsis functionality in shape and dims

ricardoV94 committed:
Allow size broadcasting from dims and observed

ricardoV94 committed:
Allow to specify dims on the fly from observed

ricardoV94 committed:
Fix _make_nice_attr_error_ for random and add it to SymbolicDistributions

ricardoV94 commented on pymc/tests/test_shape_handling.py:
```diff
@@ -304,7 +324,7 @@ def test_can_resize_data_defined_size(self):
             x = pm.MutableData("x", [[1, 2, 3, 4]], dims=("first", "second"))
             y = pm.Normal("y", mu=0, dims=("first", "second"))
-            z = pm.Normal("z", mu=y, observed=np.ones((1, 4)))
+            z = pm.Normal("z", mu=y, observed=np.ones((1, 4)), size=y.shape)
```
This behavior was neat, but could cause bugs, because observed may broadcast the shape implied by the parameters. 

One must specify explicitly how the shape flows from the parameters as in this updated test.

michaelosthege replied:
As far as I can tell, this `z` variable doesn't need a `size` kwarg, because the shape is already implied by `mu=y`.
And if `x` was resized, the `observed` shape wouldn't matter and there would be a shape error.

Or did I overlook something?
Because if feels very wrong that observed RVs need the `shape` but unobserved don't..

ricardoV94 replied:
Observed variables don't need the shape argument, unless you want to tell PyMC something that it would otherwise not be able to know: i.e., it's shape is the same as another variable or one of it's inputs.

Otherwise you would incorrectly miss these cases:

```python
def test_broadcast_by_observed(self):
        with pm.Model() as m:
            y = pm.Normal("y")
            x = pm.Normal("x", mu=y, observed=np.zeros((3,)))
            assert x.eval().shape == (3,)
```
You can't just ignore the shape of observed. Notice that  all this is just a shortcut so the user doesn't have to specify shape when we have observed. And the correct thing is to just replace `shape=observed.shape`... unless the user gives us shape directly. 

We can't safely know where the source of the batched shape is coming from otherwise.

michaelosthege replied:
I understand the example,  but in multiple of these changes in the tests I don't see why it's necessary.

Is this the behavioral change because you are resizing the RV to the observed shape even if it already matches?
And this unlinks it from the symbolic shape that is implied by the parameters?

I'm not sure if that's a great idea..  This is a non-obvious need to specify shape in order **to turn off** the smart behavior of broadcasting to observed shape..

It would be more consistent to **not** automatically broadcast to the observed shape, but require to specify shape/size in these cases.
This would also mean that the shape is inherited from parameters just like with unobserved variables.

ricardoV94 replied:
> It would be more consistent to not automatically broadcast to the observed shape, but require to specify shape/size in these cases.
This would also mean that the shape is inherited from parameters just like with unobserved variables.

Observed is by definition the shape of the variable, wheres implied dimensions is not. I understand it's not as neat as what we had but the problem is that in Aesara we don't know if parameters are going to have to be broadcasted or not many times.

michaelosthege replied:
Taken together with the other breaks I think we should do a major version bump then.
This started as a fix of a rare edge case,  but now it is a breaking change that effects potentially many lines of user code.

michaelosthege replied:
Please edit the PR description to mention this new need for explicit shape/size!

ricardoV94 replied:
> Please edit the PR description to mention this new need for explicit shape/size!

Added this to major/breaking changes:
- When shape is not provided, dims or observed are used to define the shape of a variable. If you want a variable that has dims or is observed to automatically resize when its inputs change, you must specify that explicitly via the shape argument. For example, `pm.Normal("likelihood", mu=mu, sigma=sigma, observed=data, dims="data", shape=mu.shape)`

twiecki commented on pymc/distributions/distribution.py:
```diff
@@ -150,35 +149,35 @@ def fn(*args, **kwargs):
+    # without dims information and then resize it.
+    # TODO: Consider removing dims-on-the-fly for simplicity sake.
+    if not size_or_shape and observed is not None:
+        kwargs["shape"] = shape_from_observed(observed)
```
Does this have any effect?

ricardoV94 replied:
Yes, check the new `test_define_dims_on_the_fly_from_observed`

ricardoV94 commented:
Started a discussion for removing dims on the fly: https://github.com/pymc-devs/pymc/discussions/6065

ricardoV94 commented:
I seem to have broken some things according to the failing `test_density_scaling_with_generator`. It seems that we support generators as observed data... which means we can only inquiry about data once (and then we must pass it around). Is this something we really need?

Seems to be used by Minibatch, cc @ferrine

aseyboldt commented on pymc/distributions/distribution.py:
```diff
@@ -150,35 +149,34 @@ def fn(*args, **kwargs):
-        resize_shape, observed = resize_from_observed(observed, ndim_actual)
-    return rv_out, dims, observed, resize_shape
+
+    if not size_or_shape and dims is not None:
```
Couldn't we use this spot to do some error checking? If a size or shape is specified and dims is also specified, then we could check if they are consistent, and throw an error if they are not. So we wouldn't have to say something "shape takes precedent", but just: "If you specify both they must say the same thing". Otherwise I think we would get an error when we try to export a trace to arviz, right? And we'd rather get the error at definition, than getting it after sampling? Or is there a check for this somewhere else already?

We could then add a test like this:

```python
with pytest.raises(ValueError):
    with pm.Model(coords={"a": [1, 2]}) as model:
        pm.Normal("y", shape=3, dims="a")
```

ricardoV94 replied:
The question about precedence in this comment/ section is not related to error checking but what inputs ultimately go into the RV size argument. The reason we give preference to `size`/`shape` is because it can be an arbitrary expression and so is more powerful in principle than dims.

This PR keeps the same level of no-inconsistency checking as before. The direct tests in `test_shape_handling` mention this in their comments. 

I would leave such discussion outside of this PR unless it feels very relevant.

aseyboldt replied:
Sure, it isn't really related to the issue in the PR.

I'll open a separate PR about that, I think we can at least do some checks if the shape/size is an integer. Has been too long anyway since I did my last pymc PR. :-)

ricardoV94 committed:
Filter non-parameter inputs of RandomVariables in model_graph

This removes visual dependencies between observed data and likelihood, due to flow of shape information

ricardoV94 commented:
Tests are passing

michaelosthege commented:
> Tests are passing

👍 I will review this tonight!

ricardoV94 commented:
> > Tests are passing
> 
> +1 I will review this tonight!

I'll wait for it!

michaelosthege commented on pymc/distributions/distribution.py:
```diff
@@ -150,35 +149,34 @@ def fn(*args, **kwargs):
+    # specified on the fly, we need a two-step process where we first create the RV
+    # without dims information and then resize it.
+    if not size_or_shape and observed is not None:
+        kwargs = kwargs.copy()
```
Because it's `**kwargs` in the signature, there's no need to `.copy()`.
```suggestion
```

ricardoV94 replied:
Thanks 👍

michaelosthege commented on pymc/distributions/shape_utils.py:
```diff
@@ -490,85 +465,50 @@ def convert_size(size: Size) -> Optional[StrongSize]:
-    resize_shape: Tuple[Variable, ...] = tuple(
-        model.dim_lengths[dname] for dname in sdims[:ndim_resize]
+    return tuple(
+        model.dim_lengths[dname] if dname in model.dim_lengths else shape_implied[i]
```
```suggestion
        model.dim_lengths.get(dname, shape_implied[i])
```

ricardoV94 replied:
Can't do that because dims may extend beyond the implied shape and the fallback expression in `get` is eagerly evaluated. You would get an IndexError

michaelosthege commented on pymc/tests/test_data_container.py:
```diff
@@ -45,7 +45,7 @@ def test_sample(self):
             x_shared = pm.MutableData("x_shared", x)
             b = pm.Normal("b", 0.0, 10.0)
-            pm.Normal("obs", b * x_shared, np.sqrt(1e-2), observed=y)
+            pm.Normal("obs", b * x_shared, np.sqrt(1e-2), observed=y, shape=x_shared.shape)
```
This change is technically not needed, right?

Because the `mu` parameter here is already shape `(100,)` just like the `y`..

ricardoV94 replied:
It's needed for the same reason as above. Observed can broadcast implied dims and not the other way around. It could be that mu has shape 1.

ricardoV94 replied:
Only needed because we want to resize it indirectly through x later.

michaelosthege commented on pymc/tests/test_sampling.py:
```diff
@@ -1606,7 +1606,7 @@ def test_linear_model(self):
             mu = pm.Deterministic("mu", alpha + beta * x)
             sigma = pm.HalfNormal("sigma", 0.1)
-            obs = pm.Normal("obs", mu, sigma, observed=y)
+            obs = pm.Normal("obs", mu, sigma, observed=y, shape=x.shape)
```
Similar to above: Why pass `shape`?

Oh, is this to keep `obs` resizeable?
(See comments here: https://github.com/pymc-devs/pymc/pull/6063/files#r955056055)

ricardoV94 replied:
It's just so that we don't get an extra graph dependency and do not have to change the tests expectations. I could equally have kept the original graph and add the new expected variable in the checks below. This was faster.

michaelosthege commented on pymc/tests/test_shape_handling.py:
```diff
@@ -249,69 +247,60 @@ def test_param_and_batch_shape_combos(
+            assert "ddata" in pmodel.dim_lengths
+
+            # Note: No checks are performed that observed and dims actually match.
+            y = pm.Normal("y", observed=[0, 0, 0], dims="ddata")
+            assert pmodel.RV_dims["y"] == ("ddata",)
+            assert y.eval().shape == (3,)
```
Don't you want to broadcast in this test too?
```suggestion
            y = pm.Normal("y", observed=[[0, 0, 0]], dims=("dobs", "ddata"))
            assert pmodel.RV_dims["y"] == ("dobs", "ddata")
            assert y.eval().shape == (1, 3)
```

ricardoV94 replied:
Not sure, is this not already covered by `test_define_dims_on_the_fly_from_observed`? If it is the same I would like to keep it separate, in case we decide to drop support for dims on the fly.

michaelosthege commented on pymc/distributions/distribution.py:
```diff
@@ -258,15 +256,21 @@ def __new__(
+        if observed is not None:
+            observed = convert_observed_data(observed)
+
+        # Create the RV, without taking into consideration `dims`
```
```suggestion
        # Create the RV, without taking `dims` into consideration
```

michaelosthege commented on pymc/distributions/shape_utils.py:
```diff
@@ -490,85 +465,50 @@ def convert_size(size: Size) -> Optional[StrongSize]:
-        # We don't have a way to know the names of implied
-        # dimensions, so they will be `None`.
-        dims = (*dims[:-1], *[None] * ndim_implied)
     sdims = cast(StrongDims, dims)
```
Signature violation: The previous `dims: WeakDims` parameter was already a sequence, but with `dims: Dims` the new signature allows for `dims="bla"`.
This is not accounted for in the implementation; the `cast(StrongDims, dims)` is not safe yet.

ricardoV94 replied:
Removed the cast, now that the signature of `shape_from_dims` requires `StrongDims`

michaelosthege commented on pymc/distributions/distribution.py:
```diff
@@ -150,35 +149,34 @@ def fn(*args, **kwargs):
-    return rv_out, dims, observed, resize_shape
+
+    if not size_or_shape and dims is not None:
+        resize_shape_from_dims = shape_from_dims(dims, tuple(rv_out.shape), model)
```
This call has `dims: StrongDims` which means that `shape_from_dims` can be typed to require `dims: StrongDims`

ricardoV94 replied:
done

michaelosthege commented on pymc/distributions/distribution.py:
```diff
@@ -460,19 +452,21 @@ def __new__(
-                rv=rv_out,
-                new_size=resize_shape,
-                expand=True,
+        # Resize variable based on `dims` information
+        if resize_shape_from_dims:
```
Why is this not covered?

ricardoV94 replied:
We didn't have a case with SymbolicDistributions. I am removing this experimental Class in the follow up PR I already opened anyway, so I didn't want to add nearly duplicated tests just for this.

michaelosthege replied:
Ah I didn't notice that this was in `SymbolicDistribution`

ricardoV94 commented:
@michaelosthege Do you see a difference between broadcasting by dims and by observed? Because all the same changes would have been needed if the tests had used dims instead of observed and then assumed you could just resize a variable indirectly by resizing it's arguments.

michaelosthege commented:
> @michaelosthege Do you see a difference between broadcasting by dims and by observed? Because all the same changes would have been needed if the tests had used dims instead of observed and then assumed you could just resize a variable indirectly by resizing it's arguments.

I see you point, but I think for most people the intuition is that shape, size and dims are an input of the RV, whereas the observed is something that the RV points to.
That's also how we draw it in `pm.model_to_graphviz` and it resembles the direction of the generative process.

That's probably why I don't like to broadcast automatically based on `observed`. (I found that strange and complicated when I worked on these implementations in the past..)

ricardoV94 commented:
The graphviz is a bit strange because we use 2 nodes (RV and Value) to represent what is done by a single one for unobserved RVs.

Anyway, I think:
1) The output of `test_broadcast_by_observed` would be very surprising before this PR. Also would be very surprising if we ignored extra dimensions implied by observed (which we don't)
2) We should make it clear (in the dimensionality notebook?) that in the generative graph, observed is used as a shortcut for `shape=observed.shape` unless dims or shape is provided. I hope this provides a good mental picture so users don't expect indirect resizing by default.
3) We should make it clear that the shape argument is more powerful than before. It can take arbitrary symbolic expressions. Hopefully it will become intuitive to specify `shape=mu.shape` or whatever as we start adding it to official examples.

I agree it's more cumbersome, but it does push towards being more explicit about shapes, which we are already doing anyway when we suggest users use dims.

I agree with making it a ~~minor~~ bigger (what's the name for intermediate?) release.

ricardoV94 committed:
Add note about the effects of dims and observed on the shape of a variable

michaelosthege reviewed (approved):
I can't review super-thoroughly right now, but I trust that @ricardoV94 addressed my concerns.

I'm still not super happy about prioritizing `observed.shape` over the implied dimensionality, but we'll see how the users take it.
I know enough about PyMC so my own models don't blow up because of it ¯\_(ツ)_/¯
